pkg/reconciler/storage/storage.go

```diff
@@ -129,6 +128,17 @@ func (c *Reconciler) Reconcile(ctx context.Context, key string) error {
 }
 
 func (c *Reconciler) reconcile(ctx context.Context, csr *v1alpha1.Storage) error {
+	// spec.project is optional, so if it is not specified, find the correct value and use it,
+	// rather than spec.project for the entire reconciliation.
+	project := csr.Spec.Project
+	if project == "" {
+		var err error
+		project, err = metadata.ProjectID()
```

Comment:
Sorry this will still not work, the secrets tied to the resource need to be used to ask the correct metadata server. This will find out the gke project, not the secrets project. Ville is working on a migration to the jobs and will be PRed soon

Reply:
My understanding is indeed that this will find the project that the Cluster belongs to. The credential (from the secret) is in the json that it was created in. I think it's fine if we use this project, I could also see having to pull it off mounted secret. If the other parts use the cluster project, then we should just be consistent.

Reply:
Isn't this how the Topic operation works?

https://github.com/google/knative-gcp/blob/e0b81bfb474052be268ac900533095bcca38df8d/pkg/operations/pubsub/topic.go#L95

But I agree that we should create one function that everything in this repo uses, so that we consistently choose projects (e.g. spec.project -> secret's project -> metadata server's project).

Reply:
Yes, I guess i worded poorly. What I meant to say that I can see that's how we do it today, what I was trying to say is that if we're now thinking that perhaps we should use the project the secret was created, then we should do it everywhere.
I think it's fine to do it this way now, it's consistent with other usages in this repo, so let's do it.

Reply:
this happens inside the job using the provided secret, see the path:

`knative-gcp/pkg/operations/pubsub/topic.go`

It should not happen in the reconciler. You will get the wrong project id.

Reply:
> this happens inside the job using the provided secret, see the path:
> 
> `knative-gcp/pkg/operations/pubsub/topic.go`
> 
> It should not happen in the reconciler. You will get the wrong project id.

I completely agree that it happens inside the Job, not the reconciler. But I don't see anything in that Job code that looks at the secret to get the project. Nor do I see anything in the docs for the metadata library that says it looks at anything but the instance's metadata server. Does the use of the secret somehow modify the values in the metadata server?

Reply:
It is a pending feature. The goal is to keep spec.projectId blank and use the metadata server, and use the code Ville just added to get the resolved project id from the job and save it in status.projectId. I did not have a good way to get that info so it was left as a todo.